stl/inc/xmemory

```diff
@@ -677,35 +741,76 @@ struct _Default_allocator_traits { // traits for std::allocator
+        if (_STD is_constant_evaluated()) {
+            return _Al.allocate(_Count);
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            (void) _Al;
+            return static_cast<pointer>(
+                _Allocate<_New_alignof<value_type>>(_Get_size_of_n<sizeof(value_type)>(_Count)));
+        }
     }
 
-    _NODISCARD static __declspec(allocator) pointer
-        allocate(_Alloc&, _CRT_GUARDOVERFLOW const size_type _Count, const_void_pointer) {
-        return static_cast<pointer>(_Allocate<_New_alignof<value_type>>(_Get_size_of_n<sizeof(value_type)>(_Count)));
+    _NODISCARD static _CONSTEXPR20_DYNALLOC __declspec(allocator) pointer
+        allocate(_Alloc& _Al, _CRT_GUARDOVERFLOW const size_type _Count, const_void_pointer) {
+#ifdef __cpp_lib_is_constant_evaluated // TRANSITION, GH-1532
+        if (_STD is_constant_evaluated()) {
+            return _Al.allocate(_Count);
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            (void) _Al;
+            return static_cast<pointer>(
+                _Allocate<_New_alignof<value_type>>(_Get_size_of_n<sizeof(value_type)>(_Count)));
+        }
     }
 
-    static void deallocate(_Alloc&, const pointer _Ptr, const size_type _Count) {
+    static _CONSTEXPR20_DYNALLOC void deallocate(_Alloc& _Al, const pointer _Ptr, const size_type _Count) {
         // no overflow check on the following multiply; we assume _Allocate did that check
-        _Deallocate<_New_alignof<value_type>>(_Ptr, sizeof(value_type) * _Count);
+#ifdef __cpp_lib_is_constant_evaluated // TRANSITION, GH-1532
+        if (_STD is_constant_evaluated()) {
+            return _Al.deallocate(_Ptr, _Count);
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            (void) _Al;
+            _Deallocate<_New_alignof<value_type>>(_Ptr, sizeof(value_type) * _Count);
+        }
     }
 
     template <class _Objty, class... _Types>
-    static void construct(_Alloc&, _Objty* const _Ptr, _Types&&... _Args) {
-        ::new (const_cast<void*>(static_cast<const volatile void*>(_Ptr))) _Objty(_STD forward<_Types>(_Args)...);
+    static _CONSTEXPR20_DYNALLOC void construct(_Alloc&, _Objty* const _Ptr, _Types&&... _Args) {
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            _STD construct_at(_Ptr, _STD forward<_Types>(_Args)...);
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            ::new (_Voidify_iter(_Ptr)) _Objty(_STD forward<_Types>(_Args)...);
+        }
     }
 
     template <class _Uty>
-    static void destroy(_Alloc&, _Uty* const _Ptr) {
-        _Ptr->~_Uty();
+    static _CONSTEXPR20_DYNALLOC void destroy(_Alloc&, _Uty* const _Ptr) {
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            _STD destroy_at(_Ptr);
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            _Ptr->~_Uty();
+        }

```

Comment:
I am in favor of using `_STD destroy_at` unconditionally to avoid code duplicates. I'll add that next commit.

Reply:
More follow up:
I am thinking of an implementation like that could work as well
```C++
#if _HAS_CXX17
    _STD destroy_at(_Ptr);
#else
    _Ptr->~Uty();
#endif
```
What do you think?

Reply:
From my reading, we are considering 2 different things that we can toggle on/off (whether the function is constant evaluated, and whether this function is called in C++20 mode); this results in 4 scenarios:

1. C++20 and constant-evaluated
2. C++20 and not constant-evaluated
3. pre-C++20 and constant-evaluated
4. pre-C++20 and not constant-evaluated

Option 3 should not be possible, as this function is not marked `constexpr` until C++20.  

Options 1 and 2 should call `_STD destroy_at` given WG21-P0784, since change in the description of `std::allocator_traits::destroy` does not hinge on whether the function is constant-evaluated or not, it simply changes that behavior in C++20.  

Option 4 should preserve pre-C++20 behavior and continue to call `_Ptr->~Uty()` as it did before.  (I was mistaken in my original comment - we can't call `_STD destroy_at` _unconditionally_, but I believe we should do it in more cases than this current implementation does).  I think we can utilize a pattern similar to what you have in your follow up comment to do this: 
```C++
#ifdef __cpp_lib_constexpr_dynamic_alloc
   _STD destroy_at(_Ptr);
#else
   _Ptr->~Uty();
#endif // __cpp_lib_constexpr_dynamic_alloc
```

Reply:
This should be resolved with commit 686b519 and 580db38